lib/Dialect/FIRRTL/Transforms/GroupMerge.cpp

```diff
@@ -0,0 +1,81 @@
+//===- GroupMerge.cpp - Merge optional groups together --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass merges group definitions which reference the same group
+// declaration.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/PatternMatch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-group-merge"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+/// A pass that merges optional groups referencing the same group declaration.
+struct GroupMerge : public GroupMergeBase<GroupMerge> {
+  void runOnOperation() override;
+};
+} // namespace
+
+void GroupMerge::runOnOperation() {
+  LLVM_DEBUG(
+      llvm::dbgs() << "==----- Running GroupMerge "
+                      "--------------------------------------------------===\n"
+                   << "Module: '" << getOperation().getName() << "'\n";);
+
+  // Track the last group that we saw which referenced a specific group
+  // declaration.  Because this pass operates as a single walk of the IR, it is
+  // only ever possible that there is one prior group that references a given
+  // group declaration.
+  llvm::DenseMap<SymbolRefAttr, GroupOp> priorGroups;
+
+  // Recursively walk GroupOps in the module.  Whenever we see a group, check to
+  // see if there is a prior gropu that references the same declaration.  If
+  // not, this group becomes the prior group and we continue.  If there is a
+  // prior group, then splice the prior group's body into the beginning of this
+  // group and erase the prior group.  This group then becomes the new prior
+  // group.
+  //
+  // The recursive walk will cause nested groups to also be merged.
+  auto moduleOp = getOperation();
+  mlir::IRRewriter rewriter(moduleOp.getContext());
+  moduleOp.walk([&](Operation *op) {
```

Comment:
nit: Maybe use group op? Probably the intension is to skip nested regions (even though there is almost no nest regions  after ExpandWhens) but the default order of Operation::walk is post-order so `WalkResult::skip` doesn't actually do micro optimization. 
```suggestion
  moduleOp.walk([&](GroupOp op) {
```

Reply:
👍 The intent was to visit all `GroupOp`s and recurse into any `GroupOp`s found. The `GroupOp`s retain their nesting after `ExpandWhens`. I had just forgotten that you can use `walk(GroupOp ...)` to only visit what you want.

The DFS post-order is also fine here. It may temporarily create invalid IR as nested groups are merged before their parent groups (if a nested group references a value defined in its parent, but it is merged into a different parent before its old and new parent are merged). This is a-okay, though.